### PR TITLE
Document supplement

### PR DESCRIPTION
**Pull request by dota17**
### major change:
~~rename exception type `ValueError` to `IndexError`, related issue #1119~~
+ document exception type in docs/high/dataset.rst, related issue #1447 

~~add some note when creating ragged arrays, related issue #1548~~

CC: @takluyver

### CODE REVIEW

File: docs/special.rst [hunk 1]
```diff
@@ -91,6 +91,11 @@ For example, you can create a "ragged" array of integers::
     >>> dset[0] = [1,2,3]
     >>> dset[1] = [1,2,3,4,5]
 
+But unfortunately, NumPy in particular has no real support for "ragged" array, so
+everything is workarounds. The `awkward-array <https://github.com/scikit-hep/awkward-array>`_
```
Comment (takluyver) [inline docs/special.rst:95]: I think this is some of my words from an issue? I was writing a bit casually; I might need to have another go at it.

I'm also not sure whether to point people to awkward-array. I thought it allowed reading and writing ragged arrays as HDF5 vlen data, but it actually has its own way of saving its arrays in an HDF5 file, and it's only useful if you use the same tool to read it.

Comment (dota17) [reply docs/special.rst:95]: Yes, it's from https://github.com/h5py/h5py/issues/1548.
Then, I will  wait for your further suggestion.

File: h5py/h5g.pyx [hunk 2]
```diff
@@ -406,7 +406,7 @@ cdef class GroupID(ObjectID):
         Retrieve a copy of the group creation property list used to
         create this group.
         """
-        return PropGCID(H5Gget_create_plist(self.id))
```
Comment (aragilar) [inline h5py/h5g.pyx:409]: Why the change from PropGCID to propwrap?

Comment (dota17) [reply h5py/h5g.pyx:409]: because I found that the `get_create_plist` in h5d, h5f, h5t do the same thing. 😃

File: docs/high/dataset.rst [hunk 3]
```diff
@@ -420,7 +420,7 @@ dataset::
 
 The dtype of the dataset can be accessed via ``<dset>.dtype`` as per normal.
 As empty datasets cannot be sliced, some methods of datasets such as
-``read_direct`` will raise an exception if used on a empty dataset.
+``read_direct`` will raise an **TypeError** exception if used on a empty dataset.
```
Comment (tacaswell) [inline docs/high/dataset.rst:423]: Maybe use `TypeError` instead which if intersphinx is set up correctly will link to the cpython docs.

Comment (dota17) [reply docs/high/dataset.rst:423]: do you mean link `TypeError` to the cpthon docs?

Comment (aragilar) [reply docs/high/dataset.rst:423]: @dota17 
```
``TypeError``
```
or
```
`TypeError`
```
are what you want to replace `**TypeError**` with (I think the first, but maybe the second, I can never remember which one).

Comment (takluyver) [reply docs/high/dataset.rst:423]: The first one (double backticks) is an inline code snippet in rst. I'd use that here.

@tacaswell was suggesting that we use a role to link to the definition of `TypeError` in the Python docs. But we don't have intersphinx configured to link to the Python docs at the moment, and I don't think the [description of TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) is that useful to us anyway.

File: h5py/_hl/group.py [hunk 4]
```diff
@@ -475,10 +475,11 @@ def copy(self, source, dest, name=None,
        Example:
 
         >>> f = File('myfile.hdf5')
```
Comment (takluyver) [inline h5py/_hl/group.py:477]:
```suggestion
        >>> f = File('myfile.hdf5', 'w')
```

Like in the example you modified.

File: docs/high/dataset.rst (hunk 3 above)
Comment (takluyver) [inline docs/high/dataset.rst:423]:
```suggestion
``read_direct`` will raise a ``TypeError`` exception if used on a empty dataset.
```

File: h5py/h5g.pyx (hunk 2 above)
Comment (takluyver) [reply h5py/h5g.pyx:409]: That's reasonable enough. For future reference, it's a good idea to keep pull requests focused, so a documentation PR doesn't change code at all (except code in docs & examples). But this instance is OK.
